hnn_core/cell.py

```diff
@@ -51,30 +51,46 @@ def __init__(self, event_times, threshold):
         self.nrn_netcon = h.NetCon(self.nrn_vecstim, None)
         self.nrn_netcon.threshold = threshold
 
+        self._assigned_gid = None
+        if gid is not None:
+            self.gid = gid
+
+    @property
+    def gid(self):
+        return self._assigned_gid
+
+    @gid.setter
+    def gid(self, gid):
```

Comment:
is there a realistic usecase for being able to update cell gids? Otherwise, I would just hide the functionality. The less you expose, the better

Reply:
Well this is a little stronger than hiding, no? I'm not sure the `Cell` gid ever needs to be updated, certainly not by users! Once it gets set, it should raise all hell if anything tries to change it. However, it would be nice to be able to _get_ the value using `cell.gid`, hence the `@property`. I know this complicates the `_Cell` class a little, but may make reading downstream code clearer (see `network_builder:L431-440`)?

Reply:
Hmm, I'm not sure why we need 2 gid attributes here (i.e., `_assigned_gid` and `gid`). Why can't we just have one attribute named `_ArtificialCell._gid` that is set to the arg `gid` using your decorated setter function, and is retrieved by your decorated property function? Unless I'm missing something nuanced in the logic, such a change would simplify the gid code in this class as well as subsequent cell classes.

Reply:
ah, maybe a little convoluted there yes. I just pushing something now, which I instantly regretted :( Ignore first push, my proposal will be the second one! To answer your question: I'm proposing to protect `.gid` because we don't want to risk anything messing with a `Cell`'a gid once building has begun. Perhaps there's a better way of achieving this?

Reply:
What I'm imagining is something like

```python
def __init__(self, event_times, threshold, gid=None):
   ...
   self.gid = gid

@gid.setter
def gid(self, gid)
   if self.gid is None:
      self.gid = gid
   else:
      raise RuntimeError('Global ID for this cell already assigned!')
```

Reply:
well, I'm not sure it's getting any better (7d16670)? If not, could you explain your thoughts further? I like the idea of a `Cell` sporting its `gid` for downstream use. There's some duplication in the code now, though, so if you agree on something like this,  might have to create a superclass for `_Cell` and `_ArtificialCell` to inherit from?

Reply:
Oh sorry, I didn't refresh my browser and made my previous comment before seeing your most recent commit. I'm just trying to eliminate the need for having two attributes that store the gid in order to make the logic more readable. Correct me if I'm wrong, but the only consideration here is that the gid of a given cell should only be assigned to a value other than `None` once. Since this logic can be contained purely within a gid setter function, I don't understand why we need a `self.gid` and `self._gid` attribute.

Reply:
Same here with the browser refresh ;) I'm trying your proposal, but `test_cell` hangs? I would be surprised if you could do `self.gid = gid`, isn't that trying to call the 'getter' and setter at the same time?! I've seen the `_`-prefix in cases like this elsewhere, thought it as a `pattern`...

Reply:
> Correct me if I'm wrong, but the only consideration here is that the gid of a given cell should only be assigned to a value other than `None` once.

That's my goal, yes.

Reply:
let me shed some light here. You can use private attributes for stuff that you don't want the user to mess with. For example if the object is complex and updating one attribute actually affects the rest of the object. For instance if the object had an attribute called `sfreq` (sampling frequency) and another attribute called `times`, you would have to update both simultaneously. This is why it sometimes makes sense to protect certain attributes and update them only through special functions.

Reply:
> Same here with the browser refresh ;) I'm trying your proposal, but `test_cell` hangs? I would be surprised if you could do `self.gid = gid`, isn't that trying to call the 'getter' and setter at the same time?!

Shoot, you're right. I think your implementation in 7d16670 is our best bet.

Reply:
Can always be changed to something smoother later.